File: packages/utils/src/work-item-filters/configs/filters/milestone.ts
/**
 * SPDX-FileCopyrightText: 2023-present Plane Software, Inc.
 * SPDX-License-Identifier: LicenseRef-Plane-Commercial
 *
 * Licensed under the Plane Commercial License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://plane.so/legals/eula
 *
 * DO NOT remove or modify this notice.
 * NOTICE: Proprietary and confidential. Unauthorized use or distribution is prohibited.
 */

// plane imports
import type { IMilestoneInstance, TFilterProperty, TSupportedOperators } from "@plane/types";
import { EQUALITY_OPERATOR, COLLECTION_OPERATOR } from "@plane/types";
// local imports
import type { TCreateFilterConfigParams, IFilterIconConfig, TCreateFilterConfig } from "../../../rich-filters";
import {
  createFilterConfig,
  getMultiSelectConfig,
  createOperatorConfigEntry,
  getIsNullOperatorConfigEntry,
} from "../../../rich-filters";

/**
 * Milestone filter specific params
 */
export type TCreateMilestoneFilterParams = TCreateFilterConfigParams &
  IFilterIconConfig<IMilestoneInstance> & {
    milestones: IMilestoneInstance[];
  };

/**
 * Helper to get the milestone multi select config
 * @param params - The filter params
 * @returns The milestone multi select config
 */
export const getMilestoneMultiSelectConfig = (
  params: TCreateMilestoneFilterParams,
  singleValueOperator: TSupportedOperators
) =>
  getMultiSelectConfig<IMilestoneInstance, string, IMilestoneInstance>(
    {
      items: params.milestones.filter(Boolean).filter((milestone) => milestone.id && milestone.title),
      getId: (milestone) => milestone.id,
      getLabel: (milestone) => milestone.title,
      getValue: (milestone) => milestone.id,
      getIconData: (milestone) => milestone,
    },
    {
      singleValueOperator,
      ...params,
    },
    {
      ...params,
    }
  );

/**
 * Get the milestone filter config
 * @template K - The filter key
 * @param key - The filter key to use
 * @returns A function that takes parameters and returns the milestone filter config
 */
export const getMilestoneFilterConfig =
  <P extends TFilterProperty>(key: P): TCreateFilterConfig<P, TCreateMilestoneFilterParams> =>
  (params: TCreateMilestoneFilterParams) =>
    createFilterConfig<P>({
      id: key,
      label: "Milestone",
      ...params,
      icon: params.filterIcon,
      supportedOperatorConfigsMap: new Map([
        createOperatorConfigEntry(COLLECTION_OPERATOR.IN, params, (updatedParams) =>
          getMilestoneMultiSelectConfig(updatedParams, EQUALITY_OPERATOR.EXACT)
        ),
        getIsNullOperatorConfigEntry(params),
      ]),
    });

File: apps/api/plane/ee/urls/app/milestone.py
# SPDX-FileCopyrightText: 2023-present Plane Software, Inc.
# SPDX-License-Identifier: LicenseRef-Plane-Commercial
#
# Licensed under the Plane Commercial License (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# https://plane.so/legals/eula
#
# DO NOT remove or modify this notice.
# NOTICE: Proprietary and confidential. Unauthorized use or distribution is prohibited.

from django.urls import path

from plane.ee.views.app import (
    MilestoneViewSet,
    MilestoneWorkItemsEndpoint,
    MilestoneWorkItemsSearchEndpoint,
    WorkItemMilestoneEndpoint,
)

urlpatterns = [
    path(
        "workspaces/<str:slug>/projects/<uuid:project_id>/milestones/",
        MilestoneViewSet.as_view({"get": "list", "post": "create"}),
        name="project-milestones",
    ),
    path(
        "workspaces/<str:slug>/projects/<uuid:project_id>/milestones/<uuid:milestone_id>/work-items/",
        MilestoneWorkItemsEndpoint.as_view(),
        name="project-milestone-work-items",
    ),
    path(
        "workspaces/<str:slug>/projects/<uuid:project_id>/milestones/work-items/search/",
        MilestoneWorkItemsSearchEndpoint.as_view(),
        name="project-milestone-work-items-search",
    ),
    path(
        "workspaces/<str:slug>/projects/<uuid:project_id>/work-items/<uuid:work_item_id>/milestone/",
        WorkItemMilestoneEndpoint.as_view(),
        name="project-milestone-update-work-item-milestone",
    ),
    path(
        "workspaces/<str:slug>/projects/<uuid:project_id>/milestones/<uuid:pk>/",
        MilestoneViewSet.as_view(
            {
                "patch": "partial_update",
                "delete": "destroy",
            }
        ),
        name="project-milestone",
    ),
]
